[Quant][fx] Add lowering for functional linear

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72855

Summary: This adds functionality to lower reference models
involving functional linear in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_linear

Reviewers: jerryzh168

Subscribers: jerryzh168

Tasks: T107568666

Differential Revision: [D34514127](https://our.internmc.facebook.com/intern/diff/D34514127)

andrewor14 committed:
[Quant][fx] Add lowering for functional linear

Summary: This adds functionality to lower reference models
involving functional linear in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_linear

Reviewers: jerryzh168

Subscribers: jerryzh168

Tasks: T107568666

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Add lowering for functional linear"


Summary: This adds functionality to lower reference models
involving functional linear in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_linear

Reviewers: jerryzh168

Subscribers: jerryzh168

Tasks: T107568666

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Add lowering for functional linear"


Summary: This adds functionality to lower reference models
involving functional linear in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_linear

Reviewers: jerryzh168

Subscribers: jerryzh168

Tasks: T107568666

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Add lowering for functional linear"


Summary: This adds functionality to lower reference models
involving functional linear in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_linear

Reviewers: jerryzh168

Subscribers: jerryzh168

Tasks: T107568666

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Add lowering for functional linear"


Summary: This adds functionality to lower reference models
involving functional linear in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_linear

Reviewers: jerryzh168

Subscribers: jerryzh168

Tasks: T107568666

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Add lowering for functional linear"


Summary: This adds functionality to lower reference models
involving functional linear in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_linear

Reviewers: jerryzh168

Subscribers: jerryzh168

Tasks: T107568666

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Add lowering for functional linear"


Summary: This adds functionality to lower reference models
involving functional linear in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_linear

Reviewers: jerryzh168

Subscribers: jerryzh168

Tasks: T107568666

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx][improvement] Add lowering for functional linear"


Summary: This adds functionality to lower reference models
involving functional linear in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_linear

Reviewers: jerryzh168

Subscribers: jerryzh168

Tasks: T107568666

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -130,6 +139,86 @@ def _lower_weighted_ref_module(model: QuantizedGraphModule) -> QuantizedGraphMod
+            if is_relu and q_relu_func is None:
+                continue
+
+            # Set up match pattern: (dequantize - [relu_op - ] func_op - quantize)
+            # Func args: (dequantized inputs, dequantized weights[, bias])
+            # Quantize args: (func, scale, zp, dtype)
+            func_pattern: Tuple[Any, ...] = (ref_func, "dequantize", "dequantize")
+            if has_bias:
+                func_pattern = tuple(list(func_pattern) + [MatchAllNode])
+            if is_relu:
+                func_pattern = (F.relu, func_pattern)
+            pattern = (torch.quantize_per_tensor, func_pattern, MatchAllNode, MatchAllNode, MatchAllNode)
```
can we just write down the patterns?

jerryzh168 replied:
I mean we can write down the `func_pattern` part, instead of generating it

andrewor14 replied:
Ok, since there are 4 configurations do you mean having something like this:
```
if has_bias and is_relu:
    func_pattern = ...
elif has_bias and not is_relu:
    func_pattern = ...
elif not has_bias and is_relu:
    func_pattern = ...
else:
    func_pattern = ...
```

jerryzh168 replied:
can we have something similar to https://github.com/pytorch/pytorch/pull/72953/files

jerryzh168 replied:
we can generate the easy ones with simple code, and for the ones that we can't easily generate, we can write down the pattern directly

andrewor14 replied:
I wasn't able to follow https://github.com/pytorch/pytorch/pull/72953 exactly because here we need the pattern to match the configuration (e.g. `is_relu` + `has_bias`) in the loop. I did write out the `func_pattern` part more explicitly

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -130,6 +139,86 @@ def _lower_weighted_ref_module(model: QuantizedGraphModule) -> QuantizedGraphMod
+                model.graph.erase_node(func_node)
+
+                # Step 3: TODO(andrew) Fold weights
+            model.recompile()
```
nit: let's remove this for now, we can do one recompile in the end I think

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -130,6 +139,86 @@ def _lower_weighted_ref_module(model: QuantizedGraphModule) -> QuantizedGraphMod
+                    model.graph.erase_node(relu_node)
+                model.graph.erase_node(func_node)
+
+                # Step 3: TODO(andrew) Fold weights
```
yeah we just need to swap the order of weight folding and lowering: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/convert.py#L648, also we can move fold_weight to _lower_to_native_backend

jerryzh168 replied:
can we verify that folding happens? if all tests passes now we probably haven't do that, maybe we can add a test for it

andrewor14 replied:
Ok, I added a check at the end of `test_functional_linear`

jerryzh168 reviewed (approved):
looks good! had a few nit comments inline

andrewor14 committed:
Update on "[Quant][fx][improvement] Add lowering for functional linear"


Summary: This adds functionality to lower reference models
involving functional linear in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_linear

Reviewers: jerryzh168

Subscribers: jerryzh168

Tasks: T107568666

[ghstack-poisoned]

andrewor14 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -63,6 +73,77 @@ def is_fixed_qparams_node(node, modules):
+            packed_weight = packed_weights[node.name]
+            # add a prepacked attribute to root
+            op_node = list(prepack_node.users)[0]
+            module_path, _ = node_name_to_scope.get(op_node.name, ('', None))
```
This function was moved from `fx/convert.py` and this was the only line I changed. Previously it was
```
module_path, _ = node_name_to_scope.get[op_node.name]
```
but this gave me a key error for `linear_relu` so I added a default

andrewor14 committed:
Update on "[Quant][fx] Add lowering for functional linear"


Summary: This adds functionality to lower reference models
involving functional linear in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_linear

Reviewers: jerryzh168

Subscribers: jerryzh168

Tasks: T107568666

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Add lowering for functional linear"


Summary: This adds functionality to lower reference models
involving functional linear in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_linear

Reviewers: jerryzh168

Subscribers: jerryzh168

Tasks: T107568666

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Add lowering for functional linear"


Summary: This adds functionality to lower reference models
involving functional linear in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_linear

Reviewers: jerryzh168

Subscribers: jerryzh168

Tasks: T107568666

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Add lowering for functional linear"


Summary: This adds functionality to lower reference models
involving functional linear in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_linear

Reviewers: jerryzh168

Subscribers: jerryzh168

Tasks: T107568666

[ghstack-poisoned]

andrewor14 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -2307,9 +2307,9 @@ def forward(self, x):
-                "mods2_scale_0", "mods2_zero_point_0"]:
+                "mods1_0_output_scale_0", "mods1_0_output_zero_point_0",
+                "mods1_1_output_scale_0", "mods1_1_output_zero_point_0",
+                "mods2_output_scale_0", "mods2_output_zero_point_0"]:
```
Note: This PR currently breaks backward compatibility since some scale and zp attributes currently have an extra `_output` prefix from [`quantize_node`](https://github.com/pytorch/pytorch/blob/d943a3e5a85b3ce4dbbadc2f3ac1cf4aa29d597b/torch/ao/quantization/fx/utils.py#L176), since the reference pattern code [calls that](https://github.com/pytorch/pytorch/blob/d943a3e5a85b3ce4dbbadc2f3ac1cf4aa29d597b/torch/ao/quantization/fx/quantization_patterns.py#L1065) and we now take this code path.

andrewor14 replied:
(These changes are reverted. This PR no longer breaks BC)

andrewor14 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -63,6 +72,77 @@ def is_fixed_qparams_node(node, modules):
+    torch._ops.ops.quantized.conv3d_prepack,
+}
+
+def fold_weight(
```
This was moved verbatim from `fx/convert.py`

andrewor14 committed:
Update on "[Quant][fx] Add lowering for functional linear"


Summary: This adds functionality to lower reference models
involving functional linear in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_linear

Reviewers: jerryzh168

Subscribers: jerryzh168

Tasks: T107568666

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Add lowering for functional linear"


Summary: This adds functionality to lower reference models
involving functional linear in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_linear

Reviewers: jerryzh168

Subscribers: jerryzh168

Tasks: T107568666

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Add lowering for functional linear"


Summary: This adds functionality to lower reference models
involving functional linear in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_linear

Reviewers: jerryzh168

Subscribers: jerryzh168

Tasks: T107568666

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34514127).

andrewor14 committed:
Update on "[Quant][fx] Add lowering for functional linear"


Summary: This adds functionality to lower reference models
involving functional linear in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_linear

Reviewers: jerryzh168

Subscribers: jerryzh168

Tasks: T107568666

Differential Revision: [D34514127](https://our.internmc.facebook.com/intern/diff/D34514127)

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Add lowering for functional linear"


Summary: This adds functionality to lower reference models
involving functional linear in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_linear

Reviewers: jerryzh168

Subscribers: jerryzh168

Tasks: T107568666

Differential Revision: [D34514127](https://our.internmc.facebook.com/intern/diff/D34514127)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34514127).

jerryzh168 commented on torch/ao/quantization/fx/quantization_patterns.py:
```diff
@@ -974,7 +975,8 @@ def convert(self,
                         node_name_to_scope,
-                        is_input=False)
+                        is_input=False,
+                        output_prefix="")
```
I feel we might need to break BC soon since we will not be using this code

andrewor14 replied:
Ok, we can do that in a separate PR
